File: api/routers/safety.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from database import get_db
from schemas import (
    MemberHealthRecordCreate,
    MemberHealthRecordUpdate,
    MemberHealthRecordResponse,
    MemberWithHealthResponse,
    TrainingSafetyChecklistCreate,
    TrainingSafetyChecklistUpdate,
    TrainingSafetyChecklistResponse,
    EmergencyIncidentCreate,
    EmergencyIncidentUpdate,
    EmergencyIncidentResponse,
    RiskMemberCreate,
    RiskMemberUpdate,
    RiskMemberResponse,
    RiskAssessmentResponse,
    VenueHazardCreate,
    VenueHazardUpdate,
    VenueHazardResponse,
    SafetyStatOverview,
    IncidentTypeStat,
    SafetyStatItem,
    HighRiskMemberItem,
    HazardTypeStat,
    EmergencyResponseStat,
)
import services

router = APIRouter(prefix="/api/safety", tags=["safety"])


@router.get("/health-records", response_model=list[MemberHealthRecordResponse])
def get_health_records(member_id: int | None = None, db: Session = Depends(get_db)):
    return services.get_health_records(db, member_id)


@router.get("/health-records/{record_id}", response_model=MemberHealthRecordResponse)
def get_health_record(record_id: int, db: Session = Depends(get_db)):
    result = services.get_health_record(db, record_id)
    if not result:
        raise HTTPException(status_code=404, detail="Health record not found")
    return result


@router.post("/health-records", response_model=MemberHealthRecordResponse)
def create_health_record(data: MemberHealthRecordCreate, db: Session = Depends(get_db)):
    return services.create_health_record(db, data.model_dump())


@router.put("/health-records/{record_id}", response_model=MemberHealthRecordResponse)
def update_health_record(record_id: int, data: MemberHealthRecordUpdate, db: Session = Depends(get_db)):
    result = services.update_health_record(db, record_id, data.model_dump(exclude_unset=True))
    if not result:
        raise HTTPException(status_code=404, detail="Health record not found")
    return result


@router.delete("/health-records/{record_id}")
def delete_health_record(record_id: int, db: Session = Depends(get_db)):
    success = services.delete_health_record(db, record_id)
    if not success:
        raise HTTPException(status_code=404, detail="Health record not found")
    return {"success": True}


@router.get("/members/{member_id}/health", response_model=MemberWithHealthResponse)
def get_member_with_health(member_id: int, db: Session = Depends(get_db)):
    result = services.get_member_with_health(db, member_id)
    if not result:
        raise HTTPException(status_code=404, detail="Member not found")
    return result


@router.get("/checklists", response_model=list[TrainingSafetyChecklistResponse])
def get_safety_checklists(rehearsal_id: int | None = None, db: Session = Depends(get_db)):
    return services.get_safety_checklists(db, rehearsal_id)


@router.get("/checklists/{checklist_id}", response_model=TrainingSafetyChecklistResponse)
def get_safety_checklist(checklist_id: int, db: Session = Depends(get_db)):
    result = services.get_safety_checklist(db, checklist_id)
    if not result:
        raise HTTPException(status_code=404, detail="Safety checklist not found")
    return result


@router.post("/checklists", response_model=TrainingSafetyChecklistResponse)
def create_safety_checklist(data: TrainingSafetyChecklistCreate, db: Session = Depends(get_db)):
    return services.create_safety_checklist(db, data.model_dump())


@router.put("/checklists/{checklist_id}", response_model=TrainingSafetyChecklistResponse)
def update_safety_checklist(checklist_id: int, data: TrainingSafetyChecklistUpdate, db: Session = Depends(get_db)):
    result = services.update_safety_checklist(db, checklist_id, data.model_dump(exclude_unset=True))
    if not result:
        raise HTTPException(status_code=404, detail="Safety checklist not found")
    return result


@router.delete("/checklists/{checklist_id}")
def delete_safety_checklist(checklist_id: int, db: Session = Depends(get_db)):
    success = services.delete_safety_checklist(db, checklist_id)
    if not success:
        raise HTTPException(status_code=404, detail="Safety checklist not found")
    return {"success": True}


@router.get("/checklists/{checklist_id}/risk-assessment", response_model=RiskAssessmentResponse)
def get_risk_assessment(checklist_id: int, db: Session = Depends(get_db)):
    result = services.assess_risks(db, checklist_id)
    if not result:
        raise HTTPException(status_code=404, detail="Safety checklist not found")
    return result


@router.get("/incidents", response_model=list[EmergencyIncidentResponse])
def get_emergency_incidents(checklist_id: int | None = None, member_id: int | None = None, db: Session = Depends(get_db)):
    return services.get_emergency_incidents(db, checklist_id, member_id)


@router.get("/incidents/{incident_id}", response_model=EmergencyIncidentResponse)
def get_emergency_incident(incident_id: int, db: Session = Depends(get_db)):
    result = services.get_emergency_incident(db, incident_id)
    if not result:
        raise HTTPException(status_code=404, detail="Emergency incident not found")
    return result


@router.post("/incidents", response_model=EmergencyIncidentResponse)
def create_emergency_incident(data: EmergencyIncidentCreate, db: Session = Depends(get_db)):
    return services.create_emergency_incident(db, data.model_dump())


@router.put("/incidents/{incident_id}", response_model=EmergencyIncidentResponse)
def update_emergency_incident(incident_id: int, data: EmergencyIncidentUpdate, db: Session = Depends(get_db)):
    result = services.update_emergency_incident(db, incident_id, data.model_dump(exclude_unset=True))
    if not result:
        raise HTTPException(status_code=404, detail="Emergency incident not found")
    return result


@router.post("/incidents/{incident_id}/resolve", response_model=EmergencyIncidentResponse)
def resolve_emergency_incident(incident_id: int, db: Session = Depends(get_db)):
    result = services.resolve_emergency_incident(db, incident_id)
    if not result:
        raise HTTPException(status_code=404, detail="Emergency incident not found")
    return result


@router.get("/risk-members", response_model=list[RiskMemberResponse])
def get_risk_members(checklist_id: int | None = None, member_id: int | None = None, db: Session = Depends(get_db)):
    return services.get_risk_members(db, checklist_id, member_id)


@router.get("/risk-members/{risk_member_id}", response_model=RiskMemberResponse)
def get_risk_member(risk_member_id: int, db: Session = Depends(get_db)):
    result = services.get_risk_member(db, risk_member_id)
    if not result:
        raise HTTPException(status_code=404, detail="Risk member not found")
    return result


@router.post("/risk-members", response_model=RiskMemberResponse)
def create_risk_member(data: RiskMemberCreate, db: Session = Depends(get_db)):
    return services.create_risk_member(db, data.model_dump())


@router.put("/risk-members/{risk_member_id}", response_model=RiskMemberResponse)
def update_risk_member(risk_member_id: int, data: RiskMemberUpdate, db: Session = Depends(get_db)):
    result = services.update_risk_member(db, risk_member_id, data.model_dump(exclude_unset=True))
    if not result:
        raise HTTPException(status_code=404, detail="Risk member not found")
    return result


@router.get("/hazards", response_model=list[VenueHazardResponse])
def get_venue_hazards(rehearsal_id: int | None = None, unresolved_only: bool = False, db: Session = Depends(get_db)):
    return services.get_venue_hazards(db, rehearsal_id, unresolved_only)


@router.get("/hazards/{hazard_id}", response_model=VenueHazardResponse)
def get_venue_hazard(hazard_id: int, db: Session = Depends(get_db)):
    result = services.get_venue_hazard(db, hazard_id)
    if not result:
        raise HTTPException(status_code=404, detail="Venue hazard not found")
    return result


@router.post("/hazards", response_model=VenueHazardResponse)
def create_venue_hazard(data: VenueHazardCreate, db: Session = Depends(get_db)):
    return services.create_venue_hazard(db, data.model_dump())


@router.put("/hazards/{hazard_id}", response_model=VenueHazardResponse)
def update_venue_hazard(hazard_id: int, data: VenueHazardUpdate, db: Session = Depends(get_db)):
    result = services.update_venue_hazard(db, hazard_id, data.model_dump(exclude_unset=True))
    if not result:
        raise HTTPException(status_code=404, detail="Venue hazard not found")
    return result


@router.post("/hazards/{hazard_id}/resolve", response_model=VenueHazardResponse)
def resolve_venue_hazard(hazard_id: int, db: Session = Depends(get_db)):
    result = services.resolve_venue_hazard(db, hazard_id)
    if not result:
        raise HTTPException(status_code=404, detail="Venue hazard not found")
    return result


@router.get("/stats/overview", response_model=SafetyStatOverview)
def get_safety_overview_stats(db: Session = Depends(get_db)):
    return services.get_safety_overview_stats(db)


@router.get("/stats/incident-types", response_model=list[IncidentTypeStat])
def get_incident_type_stats(db: Session = Depends(get_db)):
    return services.get_incident_type_stats(db)


@router.get("/stats/rehearsals", response_model=list[SafetyStatItem])
def get_rehearsal_safety_stats(db: Session = Depends(get_db)):
    return services.get_rehearsal_safety_stats(db)


@router.get("/stats/high-risk-members", response_model=list[HighRiskMemberItem])
def get_high_risk_members_stats(db: Session = Depends(get_db)):
    return services.get_high_risk_members_stats(db)


@router.get("/stats/hazard-types", response_model=list[HazardTypeStat])
def get_hazard_type_stats(db: Session = Depends(get_db)):
    return services.get_hazard_type_stats(db)


@router.get("/stats/emergency-response", response_model=EmergencyResponseStat)
def get_emergency_response_stats(db: Session = Depends(get_db)):
    return services.get_emergency_response_stats(db)

File: src/stores/safety.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import type {
  MemberHealthRecord,
  MemberWithHealth,
  TrainingSafetyChecklist,
  EmergencyIncident,
  RiskMember,
  RiskAssessment,
  VenueHazard,
  SafetyStatOverview,
  IncidentTypeStat,
  SafetyStatItem,
  HighRiskMemberItem,
  HazardTypeStat,
  EmergencyResponseStat,
  GroundCondition,
  WeatherCondition,
  IncidentType,
  IncidentSeverity,
  HazardType,
  RiskMemberStatus,
} from '@/types'
import { useApi } from '@/composables/useApi'

export const useSafetyStore = defineStore('safety', () => {
  const api = useApi()

  const healthRecords = ref<MemberHealthRecord[]>([])
  const memberWithHealth = ref<MemberWithHealth | null>(null)
  const checklists = ref<TrainingSafetyChecklist[]>([])
  const currentChecklist = ref<TrainingSafetyChecklist | null>(null)
  const incidents = ref<EmergencyIncident[]>([])
  const currentIncident = ref<EmergencyIncident | null>(null)
  const riskMembers = ref<RiskMember[]>([])
  const currentRiskMember = ref<RiskMember | null>(null)
  const hazards = ref<VenueHazard[]>([])
  const currentHazard = ref<VenueHazard | null>(null)
  const riskAssessment = ref<RiskAssessment | null>(null)

  const statsOverview = ref<SafetyStatOverview | null>(null)
  const incidentTypeStats = ref<IncidentTypeStat[]>([])
  const rehearsalStats = ref<SafetyStatItem[]>([])
  const highRiskMemberStats = ref<HighRiskMemberItem[]>([])
  const hazardTypeStats = ref<HazardTypeStat[]>([])
  const emergencyResponseStats = ref<EmergencyResponseStat | null>(null)

  const loading = ref(false)

  async function fetchHealthRecords(memberId?: number) {
    loading.value = true
    try {
      healthRecords.value = await api.safety.healthRecords.list(memberId)
    } catch (e) {
      console.error('Failed to fetch health records', e)
      healthRecords.value = []
    } finally {
      loading.value = false
    }
  }

  async function getHealthRecord(recordId: number) {
    try {
      return await api.safety.healthRecords.get(recordId)
    } catch (e) {
      console.error('Failed to get health record', e)
      return null
    }
  }

  async function createHealthRecord(data: {
    member_id: number
    record_date: string
    condition_type: string
    description?: string
    is_chronic?: boolean
    needs_accommodation?: boolean
    accommodation_notes?: string
  }) {
    const result = await api.safety.healthRecords.create(data)
    healthRecords.value.push(result)
    return result
  }

  async function updateHealthRecord(
    recordId: number,
    data: {
      condition_type?: string
      description?: string
      is_chronic?: boolean
      needs_accommodation?: boolean
      accommodation_notes?: string
    }
  ) {
    const result = await api.safety.healthRecords.update(recordId, data)
    const idx = healthRecords.value.findIndex((r) => r.id === recordId)
    if (idx >= 0) {
      healthRecords.value[idx] = result
    }
    return result
  }

  async function deleteHealthRecord(recordId: number) {
    await api.safety.healthRecords.delete(recordId)
    healthRecords.value = healthRecords.value.filter((r) => r.id !== recordId)
  }

  async function fetchMemberWithHealth(memberId: number) {
    try {
      memberWithHealth.value = await api.safety.getMemberWithHealth(memberId)
    } catch (e) {
      console.error('Failed to fetch member with health', e)
      memberWithHealth.value = null
    }
  }

  async function fetchChecklists(rehearsalId?: number) {
    loading.value = true
    try {
      checklists.value = await api.safety.checklists.list(rehearsalId)
    } catch (e) {
      console.error('Failed to fetch checklists', e)
      checklists.value = []
    } finally {
      loading.value = false
    }
  }

  async function getChecklist(checklistId: number) {
    loading.value = true
    try {
      currentChecklist.value = await api.safety.checklists.get(checklistId)
    } catch (e) {
      console.error('Failed to get checklist', e)
      currentChecklist.value = null
    } finally {
      loading.value = false
    }
  }

  async function createChecklist(data: {
    rehearsal_id: number
    ground_condition: GroundCondition
    ground_notes?: string
    audio_cables_arranged?: boolean
    audio_cables_notes?: string
    members_illness_reported?: boolean
    illness_notes?: string
    weather_temperature?: number
    weather_condition?: WeatherCondition
    weather_notes?: string
    drinking_water_provided?: boolean
    rest_schedule_arranged?: boolean
    rest_notes?: string
    high_risk_moves_reminded?: boolean
    high_risk_moves_notes?: string
    created_by?: number
  }) {
    const result = await api.safety.checklists.create(data)
    checklists.value.unshift(result)
    return result
  }

  async function updateChecklist(
    checklistId: number,
    data: {
      ground_condition?: GroundCondition
      ground_notes?: string
      audio_cables_arranged?: boolean
      audio_cables_notes?: string
      members_illness_reported?: boolean
      illness_notes?: string
      weather_temperature?: number
      weather_condition?: WeatherCondition
      weather_notes?: string
      drinking_water_provided?: boolean
      rest_schedule_arranged?: boolean
      rest_notes?: string
      high_risk_moves_reminded?: boolean
      high_risk_moves_notes?: string
    }
  ) {
    const result = await api.safety.checklists.update(checklistId, data)
    const idx = checklists.value.findIndex((c) => c.id === checklistId)
    if (idx >= 0) {
      checklists.value[idx] = result
    }
    if (currentChecklist.value?.id === checklistId) {
      currentChecklist.value = result
    }
    return result
  }

  async function deleteChecklist(checklistId: number) {
    await api.safety.checklists.delete(checklistId)
    checklists.value = checklists.value.filter((c) => c.id !== checklistId)
    if (currentChecklist.value?.id === checklistId) {
      currentChecklist.value = null
    }
  }

  async function assessRisks(checklistId: number) {
    loading.value = true
    try {
      riskAssessment.value = await api.safety.checklists.assessRisks(checklistId)
    } catch (e) {
      console.error('Failed to assess risks', e)
      riskAssessment.value = null
    } finally {
      loading.value = false
    }
  }

  async function fetchIncidents(checklistId?: number, memberId?: number) {
    loading.value = true
    try {
      incidents.value = await api.safety.incidents.list(checklistId, memberId)
    } catch (e) {
      console.error('Failed to fetch incidents', e)
      incidents.value = []
    } finally {
      loading.value = false
    }
  }

  async function getIncident(incidentId: number) {
    try {
      currentIncident.value = await api.safety.incidents.get(incidentId)
    } catch (e) {
      console.error('Failed to get incident', e)
      currentIncident.value = null
    }
  }

  async function createIncident(data: {
    checklist_id: number
    member_id: number
    incident_type: IncidentType
    song_id?: number
    position_id?: string
    formation_position?: string
    description?: string
    severity?: IncidentSeverity
    treatment_given?: string
    treated_by?: string
    family_notified?: boolean
    family_notification_details?: string
    community_leader_notified?: boolean
    community_notification_details?: string
    follow_up_required?: boolean
    follow_up_notes?: string
  }) {
    const result = await api.safety.incidents.create(data)
    incidents.value.unshift(result)
    return result
  }

  async function updateIncident(
    incidentId: number,
    data: {
      description?: string
      severity?: IncidentSeverity
      treatment_given?: string
      treated_by?: string
      family_notified?: boolean
      family_notification_details?: string
      community_leader_notified?: boolean
      community_notification_details?: string
      follow_up_required?: boolean
      follow_up_notes?: string
      resolved?: boolean
    }
  ) {
    const result = await api.safety.incidents.update(incidentId, data)
    const idx = incidents.value.findIndex((i) => i.id === incidentId)
    if (idx >= 0) {
      incidents.value[idx] = result
    }
    if (currentIncident.value?.id === incidentId) {
      currentIncident.value = result
    }
    return result
  }

  async function resolveIncident(incidentId: number) {
    const result = await api.safety.incidents.resolve(incidentId)
    const idx = incidents.value.findIndex((i) => i.id === incidentId)
    if (idx >= 0) {
      incidents.value[idx] = result
    }
    return result
  }

  async function fetchRiskMembers(checklistId?: number, memberId?: number) {
    loading.value = true
    try {
      riskMembers.value = await api.safety.riskMembers.list(checklistId, memberId)
    } catch (e) {
      console.error('Failed to fetch risk members', e)
      riskMembers.value = []
    } finally {
      loading.value = false
    }
  }

  async function getRiskMember(riskMemberId: number) {
    try {
      currentRiskMember.value = await api.safety.riskMembers.get(riskMemberId)
    } catch (e) {
      console.error('Failed to get risk member', e)
      currentRiskMember.value = null
    }
  }

  async function createRiskMember(data: {
    checklist_id: number
    member_id: number
    risk_level: string
    risk_factors?: string
    recommendation?: string
  }) {
    const result = await api.safety.riskMembers.create(data)
    riskMembers.value.push(result)
    return result
  }

  async function updateRiskMember(
    riskMemberId: number,
    data: {
      risk_level?: string
      risk_factors?: string
      recommendation?: string
      action_taken?: string
      status?: RiskMemberStatus
    }
  ) {
    const result = await api.safety.riskMembers.update(riskMemberId, data)
    const idx = riskMembers.value.findIndex((r) => r.id === riskMemberId)
    if (idx >= 0) {
      riskMembers.value[idx] = result
    }
    return result
  }

  async function fetchHazards(rehearsalId?: number, unresolvedOnly?: boolean) {
    loading.value = true
    try {
      hazards.value = await api.safety.hazards.list(rehearsalId, unresolvedOnly)
    } catch (e) {
      console.error('Failed to fetch hazards', e)
      hazards.value = []
    } finally {
      loading.value = false
    }
  }

  async function getHazard(hazardId: number) {
    try {
      currentHazard.value = await api.safety.hazards.get(hazardId)
    } catch (e) {
      console.error('Failed to get hazard', e)
      currentHazard.value = null
    }
  }

  async function createHazard(data: {
    rehearsal_id: number
    hazard_type: HazardType
    location?: string
    description?: string
    severity?: string
    reported_by?: number
  }) {
    const result = await api.safety.hazards.create(data)
    hazards.value.unshift(result)
    return result
  }

  async function updateHazard(
    hazardId: number,
    data: {
      hazard_type?: HazardType
      location?: string
      description?: string
      severity?: string
      resolved?: boolean
      resolution_notes?: string
    }
  ) {
    const result = await api.safety.hazards.update(hazardId, data)
    const idx = hazards.value.findIndex((h) => h.id === hazardId)
    if (idx >= 0) {
      hazards.value[idx] = result
    }
    return result
  }

  async function resolveHazard(hazardId: number) {
    const result = await api.safety.hazards.resolve(hazardId)
    const idx = hazards.value.findIndex((h) => h.id === hazardId)
    if (idx >= 0) {
      hazards.value[idx] = result
    }
    return result
  }

  async function fetchStatsOverview() {
    try {
      statsOverview.value = await api.safety.stats.overview()
    } catch (e) {
      console.error('Failed to fetch stats overview', e)
      statsOverview.value = null
    }
  }

  async function fetchIncidentTypeStats() {
    try {
      incidentTypeStats.value = await api.safety.stats.incidentTypes()
    } catch (e) {
      console.error('Failed to fetch incident type stats', e)
      incidentTypeStats.value = []
    }
  }

  async function fetchRehearsalStats() {
    try {
      rehearsalStats.value = await api.safety.stats.rehearsals()
    } catch (e) {
      console.error('Failed to fetch rehearsal stats', e)
      rehearsalStats.value = []
    }
  }

  async function fetchHighRiskMemberStats() {
    try {
      highRiskMemberStats.value = await api.safety.stats.highRiskMembers()
    } catch (e) {
      console.error('Failed to fetch high risk member stats', e)
      highRiskMemberStats.value = []
    }
  }

  async function fetchHazardTypeStats() {
    try {
      hazardTypeStats.value = await api.safety.stats.hazardTypes()
    } catch (e) {
      console.error('Failed to fetch hazard type stats', e)
      hazardTypeStats.value = []
    }
  }

  async function fetchEmergencyResponseStats() {
    try {
      emergencyResponseStats.value = await api.safety.stats.emergencyResponse()
    } catch (e) {
      console.error('Failed to fetch emergency response stats', e)
      emergencyResponseStats.value = null
    }
  }

  async function fetchAllStats() {
    loading.value = true
    try {
      await Promise.all([
        fetchStatsOverview(),
        fetchIncidentTypeStats(),
        fetchRehearsalStats(),
        fetchHighRiskMemberStats(),
        fetchHazardTypeStats(),
        fetchEmergencyResponseStats(),
      ])
    } finally {
      loading.value = false
    }
  }

  return {
    healthRecords,
    memberWithHealth,
    checklists,
    currentChecklist,
    incidents,
    currentIncident,
    riskMembers,
    currentRiskMember,
    hazards,
    currentHazard,
    riskAssessment,
    statsOverview,
    incidentTypeStats,
    rehearsalStats,
    highRiskMemberStats,
    hazardTypeStats,
    emergencyResponseStats,
    loading,
    fetchHealthRecords,
    getHealthRecord,
    createHealthRecord,
    updateHealthRecord,
    deleteHealthRecord,
    fetchMemberWithHealth,
    fetchChecklists,
    getChecklist,
    createChecklist,
    updateChecklist,
    deleteChecklist,
    assessRisks,
    fetchIncidents,
    getIncident,
    createIncident,
    updateIncident,
    resolveIncident,
    fetchRiskMembers,
    getRiskMember,
    createRiskMember,
    updateRiskMember,
    fetchHazards,
    getHazard,
    createHazard,
    updateHazard,
    resolveHazard,
    fetchStatsOverview,
    fetchIncidentTypeStats,
    fetchRehearsalStats,
    fetchHighRiskMemberStats,
    fetchHazardTypeStats,
    fetchEmergencyResponseStats,
    fetchAllStats,
  }
})
